File: client/server/requestValidation.ts
import type { InlinePreviewFiles, JsonObject } from "../src/types/apiContract";
import { API_DEFAULT_LIMITS, SOURCE_NAME_PATTERN, SUPPORTED_IMAGE_EXTENSIONS } from "../src/types/apiContract";

export class HttpError extends Error {
  constructor(
    public readonly statusCode: number,
    message: string,
    public readonly code = "bad_request",
    public readonly details: Record<string, unknown> = {},
  ) {
    super(message);
    this.name = "HttpError";
  }
}

export const MAX_REQUEST_BODY_BYTES = envLimit(
  "ASSET_PIPELINE_MAX_REQUEST_BODY_BYTES",
  API_DEFAULT_LIMITS.maxRequestBodyBytes,
);
export const MAX_SOURCES = envLimit("ASSET_PIPELINE_MAX_SOURCES", API_DEFAULT_LIMITS.maxSources);
export const MAX_WORKERS = envLimit("ASSET_PIPELINE_MAX_WORKERS", API_DEFAULT_LIMITS.maxWorkers);
export const MAX_PIPELINES = envLimit("ASSET_PIPELINE_MAX_PIPELINES", API_DEFAULT_LIMITS.maxPipelines);

export function requireString(value: unknown, field: string, maxLength = 256): string {
  if (typeof value !== "string" || value.trim().length === 0) {
    throw new HttpError(422, `Field '${field}' must be a non-empty string.`, "invalid_field");
  }
  if (value.length > maxLength) {
    throw new HttpError(422, `Field '${field}' is too long (max ${maxLength} characters).`, "invalid_field");
  }
  return value;
}

export function requireObject(value: unknown, field: string): JsonObject {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new HttpError(422, `Field '${field}' must be an object.`, "invalid_field");
  }
  return value;
}

export function decodePathComponent(value: string, field: string): string {
  try {
    return decodeURIComponent(value);
  } catch (error) {
    throw new HttpError(400, `Invalid encoded ${field}: ${String(error)}`, "invalid_path");
  }
}

export function requireSourceName(value: unknown): string {
  const source = requireString(value, "source", 255);
  if (
    !SOURCE_NAME_PATTERN.test(source) ||
    !SUPPORTED_IMAGE_EXTENSIONS.some((extension) => source.toLowerCase().endsWith(extension))
  ) {
    throw new HttpError(422, "source must be a supported image filename, not a path.", "invalid_source");
  }
  return source;
}

export function requireSourceNames(value: unknown): string[] {
  if (!Array.isArray(value)) {
    throw new HttpError(422, "sources must be an array.", "invalid_sources");
  }
  if (value.length < 1 || value.length > MAX_SOURCES) {
    throw new HttpError(422, `sources must contain between 1 and ${MAX_SOURCES} items.`, "invalid_sources");
  }
  const sources = value.map(requireSourceName);
  if (new Set(sources).size !== sources.length) {
    throw new HttpError(422, "sources cannot contain duplicates.", "invalid_sources");
  }
  return sources;
}

export function requireWorkers(value: unknown): number {
  if (typeof value !== "number" || !Number.isInteger(value) || value < 1 || value > MAX_WORKERS) {
    throw new HttpError(422, `workers must be an integer between 1 and ${MAX_WORKERS}.`, "invalid_workers");
  }
  return value;
}

export function validatePipelineId(value: unknown): void {
  if (value !== undefined && (typeof value !== "string" || value.trim().length === 0 || value.length > 120)) {
    throw new HttpError(422, "pipelineId must be a non-empty string.", "invalid_pipeline");
  }
}

export function validatePipelineIds(value: unknown): void {
  if (value === undefined) return;
  if (
    !Array.isArray(value) ||
    value.length > MAX_PIPELINES ||
    value.some((item) => typeof item !== "string" || item.trim().length === 0)
  ) {
    throw new HttpError(
      422,
      `pipelineIds must contain at most ${MAX_PIPELINES} non-empty strings.`,
      "invalid_pipelines",
    );
  }
  if (new Set(value).size !== value.length) {
    throw new HttpError(422, "pipelineIds cannot contain duplicates.", "invalid_pipelines");
  }
}

export function isJsonObject(value: unknown): value is JsonObject {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

export function isInlinePreviewFiles(value: unknown): value is InlinePreviewFiles {
  if (!isJsonObject(value)) {
    return false;
  }
  return Object.values(value).every((item) => typeof item === "string");
}

export function envLimit(name: string, fallback: number): number {
  const value = Number(process.env[name] ?? fallback);
  return Number.isInteger(value) && value > 0 ? value : fallback;
}

File: asset_pipeline/services/request_validation.py
from __future__ import annotations

import os
import re
from typing import cast

from asset_pipeline.services.contracts import (
    API_DEFAULT_LIMITS,
    SOURCE_NAME_REGEX,
    SUPPORTED_IMAGE_SUFFIXES,
    JsonObject,
)
from asset_pipeline.services.errors import ValidationError


def _env_limit(name: str, fallback: int) -> int:
    try:
        value = int(os.environ.get(name, fallback))
    except (TypeError, ValueError):
        return fallback
    return value if value > 0 else fallback


MAX_REQUEST_BODY_BYTES = _env_limit("ASSET_PIPELINE_MAX_REQUEST_BODY_BYTES", API_DEFAULT_LIMITS["maxRequestBodyBytes"])
MAX_SOURCES = _env_limit("ASSET_PIPELINE_MAX_SOURCES", API_DEFAULT_LIMITS["maxSources"])
MAX_WORKERS = _env_limit("ASSET_PIPELINE_MAX_WORKERS", API_DEFAULT_LIMITS["maxWorkers"])
MAX_PIPELINES = _env_limit("ASSET_PIPELINE_MAX_PIPELINES", API_DEFAULT_LIMITS["maxPipelines"])
SOURCE_NAME_PATTERN = re.compile(SOURCE_NAME_REGEX)


def validate_source_name(value: object) -> str:
    if not isinstance(value, str) or not SOURCE_NAME_PATTERN.fullmatch(value):
        raise ValidationError("source must be a filename, not a path")
    if not value.casefold().endswith(SUPPORTED_IMAGE_SUFFIXES):
        raise ValidationError("source must be a supported image filename")
    return value


def validate_source_names(values: object) -> list[str]:
    if not isinstance(values, list):
        raise ValidationError("sources must be an array")
    if not values:
        raise ValidationError("sources must contain at least one filename")
    if len(values) > MAX_SOURCES:
        raise ValidationError(f"sources cannot contain more than {MAX_SOURCES} items")
    result = [validate_source_name(value) for value in values]
    if len(set(result)) != len(result):
        raise ValidationError("sources cannot contain duplicates")
    return result


def validate_workers(value: object) -> int:
    try:
        workers = int(value)
    except (TypeError, ValueError) as error:
        raise ValidationError("workers must be an integer") from error
    if workers < 1 or workers > MAX_WORKERS:
        raise ValidationError(f"workers must be between 1 and {MAX_WORKERS}")
    return workers


def validate_pipeline_ids(value: object) -> list[str] | None:
    if value is None:
        return None
    if not isinstance(value, list):
        raise ValidationError("pipelineIds must be an array")
    if len(value) > MAX_PIPELINES:
        raise ValidationError(f"pipelineIds cannot contain more than {MAX_PIPELINES} items")
    if any(not isinstance(item, str) or not item.strip() for item in value):
        raise ValidationError("pipelineIds must contain non-empty strings")
    if len(set(value)) != len(value):
        raise ValidationError("pipelineIds cannot contain duplicates")
    return value


def validate_payload(value: object) -> JsonObject:
    if not isinstance(value, dict):
        raise ValidationError("request body must be a JSON object")
    options = value.get("options", {})
    if not isinstance(options, dict):
        raise ValidationError("options must be an object")
    return cast(JsonObject, value)
